### PR TITLE
Pin 1772 - Add pagination to endpoints used to list objects.

### PR DESCRIPTION
**Pull request by beetlecrunch**
This PR applies also a refactoring of the API interface.

### CODE REVIEW

File: src/main/resources/interface-specification.yml [hunk 1]
```diff
@@ -76,18 +70,15 @@ paths:
       parameters:
         - in: query
           name: search
-          required: true
           schema:
             type: string
         - in: query
           name: page
```
Comment (galales) [inline src/main/resources/interface-specification.yml:78]: suggest adding a `min` constraint to let the caller know that the first page is 1.
Maybe having the default value here would be nice (same for other `page` and `limit` parameters)

File: src/main/resources/interface-specification.yml [hunk 2]
```diff
@@ -124,19 +103,23 @@ paths:
           schema:
             type: string
             maxLength: 32
+        - in: query
+          name: page
```
Comment (galales) [inline src/main/resources/interface-specification.yml:120]: same for `min` constraint + default value

**Review by galales — APPROVED** [review]

File: src/main/resources/interface-specification.yml (hunk 1 above)
Comment (beetlecrunch) [reply src/main/resources/interface-specification.yml:78]: I put them, but the generator treated them as required fields without putting any constraints or default value.

This is without constraints:
```
def getCategories(origin: Option[String], page: Option[Int], limit: Option[Int])
```
and this with openapi constraints set:
```
def getCategories(origin: Option[String], page: Int, limit: Int)
```
On the client generation side this is the generated code with the constraints:
```
def getCategories(origin: Option[String] = None, page: Option[Int] = None, limit: Option[Int] = None)
```

I agree that putting them is very useful on the API side, but generator does work as expected.
A solution is to restore these fields as mandatory and check their limitation directly in the implementation.

File: src/main/resources/interface-specification.yml (hunk 2 above)
Comment (beetlecrunch) [reply src/main/resources/interface-specification.yml:120]: [Same](https://github.com/pagopa/interop-be-party-registry-proxy/pull/115#discussion_r960594340)

File: src/main/resources/interface-specification.yml (hunk 1 above)
Comment (beetlecrunch) [reply src/main/resources/interface-specification.yml:78]: I was missing where exactly the generator passes the defaults, so my previous comment is invalid.

File: src/main/scala/it/pagopa/interop/partyregistryproxy/api/impl/InstitutionApiServiceImpl.scala [hunk 3]
```diff
@@ -29,53 +29,49 @@ final case class InstitutionApiServiceImpl(institutionSearchService: IndexSearch
...
-      tuple => {
-        if (tuple._1.isEmpty) {
-          logger.error(s"Error while searching for institution with following search string = ${search} - Not Found")
-          searchInstitution404(problemOf(StatusCodes.NotFound, InstitutionsNotFound))
-        } else {
-          searchInstitution200(Institutions.tupled(tuple))
-        }
-      }
-    )
+    val result: Try[(List[Institution], Long)] = search match {
```
Comment (TonioGela) [inline src/main/scala/it/pagopa/interop/partyregistryproxy/api/impl/InstitutionApiServiceImpl.scala]: `.fold` 🙏

File: src/main/scala/it/pagopa/interop/partyregistryproxy/service/impl/package.scala [hunk 4]
```diff
@@ -1,20 +1,62 @@
...
 
 import javax.naming.directory.SearchResult
+import scala.reflect.ClassTag
 import scala.util.{Failure, Success, Try}
 
 package object impl {
 
+  def getItems[A: ClassTag](filters: Map[SearchField, String], page: Int, limit: Int)(
+    reader: DirectoryReader
+  )(implicit documentConverter: DocumentConverter[A]): Try[(List[A], Long)] = {
```
Comment (TonioGela) [inline src/main/scala/it/pagopa/interop/partyregistryproxy/service/impl/package.scala]: Useless parenthesis

**Review by TonioGela — APPROVED** [review]

**Pull request merged**